backend/question-service/lambda/app/scraper/handler.py

```diff
@@ -0,0 +1,189 @@
+COOLDOWN_PERIOD = int(os.getenv("COOLDOWN_PERIOD", 10 * 24 * 60 * 60)) # Default to 10 days in seconds
+
+# Set MongoDB Client and Connection as a global variable
+client = None
+collection = None
+
+def handler(event, context):
+    """Lambda function handler. Scrapes Leetcode for questions and saves them to the MongoDB database."""
+
+    # TODO: Implement custom event handling logic
+    print("The event received was: ", event)
+
+    # Establish a connection to the MongoDB server
+    global client
+    client = connect_to_mongo(MONGO_CONNECTION_STRING)
+
+    # Select the database and collection
+    db = client['questions_db']
+    global collection
+    collection = db['problems']
+
+    all_questions = parse_questions_xml()
+    questions_to_scrape = get_questions_to_scrape(all_questions)
+    problems = scrape_leetcode_from_urls(questions_to_scrape)
+    save_to_mongodb(problems)
+
+    # Close the connection
+    client.close()
+
+    return f"Successfully scraped and saved {len(problems)} questions!"
+
+def parse_questions_xml():
+    """Parse the Leetcode questions sitemap XML file and return a list of all question URLs."""
+    BASE_URL = "https://leetcode.com/sitemap/sitemap-question-1.xml"
+    response = requests.get(BASE_URL).content
+
+    warnings.filterwarnings("ignore", category=XMLParsedAsHTMLWarning)
+
+    soup = BeautifulSoup(response, "html.parser")
+    urls = soup.find_all("loc")
+    return [url.text for url in urls]
+
+def get_questions_to_scrape(urls):
+    """Return a list of question URLs to be scraped today. This is defined as all questions that have not been scraped yet, or have been updated more than 10 days ago"""
+
+    # Get all existing question URLs in the database
+    existing_urls = {question["url"]: question["updated_at"] for question in collection.find({}, {"url": 1, "updated_at": 1, "_id": False})}
+
+    # Get all question URLs that have not been scraped yet, or has not been updated within the cooldown period
+    current_time = int(time.time())
+    urls_to_scrape = [url for url in urls if url not in existing_urls or current_time - existing_urls[url] > COOLDOWN_PERIOD]
+    
+    # Set rate limit as the upper limit of questions to scrape per function invokation.
+    urls_to_scrape = urls_to_scrape[:RATE_LIMIT]
+
+    return urls_to_scrape
+
+def scrape_leetcode_from_url(url):
+    """Scrape Leetcode for a question in the URL and return question details."""
+    response = requests.get(url)
+    soup = BeautifulSoup(response.content, "html.parser")
+
+    script_tag = soup.find('script', attrs={'id': '__NEXT_DATA__'})
+
+    # Extract and parse the JSON data
+    if script_tag is not None:
+        json_data = json.loads(script_tag.string)
+        queries = json_data['props']['pageProps']['dehydratedState']['queries']
+        details = get_details_from_queries(queries) 
+        return details
+    
```

Comment:
nit: currently, if the question data cannot be found, the function 'fails' silently.

do you think if the script tag is not found/the question fails to be scraped, you can add it to an array of failed scrapes. then that can be printed out so can debug in the lambda

Reply:
Resolved in https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/commit/de1eb1590e1c47555ff72b5033e505e90f480449 as explained above.